kie-server-parent/kie-server-tests/kie-server-integ-tests-controller/src/test/java/org/kie/server/integrationtests/controller/ConcurrentModificationExceptionTest.java

```diff
@@ -0,0 +1,102 @@
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.kie.server.controller.api.model.runtime.ServerInstanceKey;
+import org.kie.server.controller.api.model.spec.ServerTemplate;
+import org.kie.server.controller.api.storage.KieServerTemplateStorage;
+import org.kie.server.controller.impl.storage.InMemoryKieServerTemplateStorage;
+
+import static org.awaitility.Awaitility.await;
+import static org.kie.server.controller.rest.ControllerUtils.marshal;
+
+@RunWith(BMUnitRunner.class)
+@BMUnitConfig( loadDirectory = "target/test-classes/bmunit", debug = true )
+public class ConcurrentModificationExceptionTest {
+
+    private KieServerTemplateStorage templateStorage = InMemoryKieServerTemplateStorage.getInstance();
+    private static final String serverTemplateId = "id";
+
+    private AtomicBoolean atomic;
+
+    @Before
+    public void setUp() {
+        ServerTemplate serverTemplate = new ServerTemplate(serverTemplateId, "name");
+        IntStream.range(0, 1000).boxed().map(i -> Integer.toString(i)).map(s -> new ServerInstanceKey(s, s, s, ""))
+                .forEach(instance -> serverTemplate.addServerInstance(instance));
+        templateStorage.store(serverTemplate);
+        atomic = new AtomicBoolean(true);
+    }
+
+    @Test(expected = Test.None.class)
+    public void testConcurrency() throws InterruptedException {
+        final ServerTemplate serverTemplate = templateStorage.load(serverTemplateId);
+
+        Thread disconnect = new Thread(() -> disconnect());
+        disconnect.start();
+
+        // Sync before marshalling
+        await().untilTrue(atomic);
+        String response = marshal("application/xml", serverTemplate);
+        System.out.println(response);
+    }
+
+    public void disconnect() {
+        ServerTemplate serverTemplate = templateStorage.load(serverTemplateId);
+
+        try {
+            atomic.set(true);
+            deleteServerTemplate(serverTemplate);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private void deleteServerTemplate(ServerTemplate serverTemplate) throws InterruptedException {
+        for (ServerInstanceKey instanceKey : serverTemplate.getServerInstanceKeys()) {
+            serverTemplate.deleteServerInstance(instanceKey.getServerInstanceId());
+            Thread.sleep(10);
+        }
+    }
+
+    @Test(expected = Test.None.class)
+    @BMScript("ConcurrentModificationException.btm")
+    public void testConcurrencyWithByteman() {
+        final ServerTemplate serverTemplate = templateStorage.load(serverTemplateId);
+
+        Thread disconnect = new Thread(() -> {
+            try {
+                deleteServerTemplate(serverTemplate);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        });
+        disconnect.start();
+
+        String response = marshal("application/xml", serverTemplate);
```

Comment:
This scenario can be overcome by overriding the marshal method to countDown the latch just before we delegate to the parent's implementation? Maybe we don't need Byteman at all.

Reply:
It have a sonar check issue,https://stackoverflow.com/questions/36283334/how-to-avoid-thread-sleep-in-unit-tests. So I think byteman is better way.

Reply:
But with CountDownLatch you don't use Thread.sleep() at all, it's a completely fine approach which is built-in in JDK. You just create a CountDownLatch and then use `await()` and `countDown()` methods.

Reply:
"with CountDownLatch you don't use Thread.sleep() at all" , can not reproduce  the issue.so ...
I create three way to test,just want to run all of them in sonar.